## Description: dix: Save touchpoint last coordinates before transform.
## Origin/Author: Yuly Novikov <ynovikov@chromium.org>
## Bug: https://bugs.freedesktop.org/show_bug.cgi?id=49347
Index: xorg-server-lts-quantal-1.13.0/dix/getevents.c
===================================================================
--- xorg-server-lts-quantal-1.13.0.orig/dix/getevents.c	2013-08-14 13:59:15.110805225 +0800
+++ xorg-server-lts-quantal-1.13.0/dix/getevents.c	2013-08-14 14:05:01.452522627 +0800
@@ -1964,32 +1964,27 @@
     default:
         return 0;
     }
-    if (t->mode == XIDirectTouch && !(flags & TOUCH_CLIENT_ID)) {
-        if (!valuator_mask_isset(&mask, 0))
-            valuator_mask_set_double(&mask, 0,
-                                     valuator_mask_get_double(touchpoint.ti->
-                                                              valuators, 0));
-        if (!valuator_mask_isset(&mask, 1))
-            valuator_mask_set_double(&mask, 1,
-                                     valuator_mask_get_double(touchpoint.ti->
-                                                              valuators, 1));
-    }
 
     /* Get our screen event co-ordinates (root_x/root_y/event_x/event_y):
      * these come from the touchpoint in Absolute mode, or the sprite in
      * Relative. */
     if (t->mode == XIDirectTouch) {
-        transformAbsolute(dev, &mask);
-
         if (!(flags & TOUCH_CLIENT_ID)) {
-            for (i = 0; i < valuator_mask_size(&mask); i++) {
+            for (i = 0; i < max(valuator_mask_size(&mask), 2); i++) {
                 double val;
 
                 if (valuator_mask_fetch_double(&mask, i, &val))
                     valuator_mask_set_double(touchpoint.ti->valuators, i, val);
+                /* If the device doesn't post new X and Y axis values,
+                 * use the last values posted.
+                 */
+                else if (i < 2 &&
+                    valuator_mask_fetch_double(touchpoint.ti->valuators, i, &val))
+                    valuator_mask_set_double(&mask, i, val);
             }
         }
 
+        transformAbsolute(dev, &mask);
         clipAbsolute(dev, &mask);
     }
     else {
Index: xorg-server-lts-quantal-1.13.0/include/inputstr.h
===================================================================
--- xorg-server-lts-quantal-1.13.0.orig/include/inputstr.h	2013-08-14 13:59:15.110805225 +0800
+++ xorg-server-lts-quantal-1.13.0/include/inputstr.h	2013-08-14 13:59:15.106805195 +0800
@@ -331,7 +331,7 @@
     uint32_t ddx_id;            /* touch ID given by the DDX */
     Bool emulate_pointer;
 
-    ValuatorMask *valuators;    /* last recorded axis values */
+    ValuatorMask *valuators;    /* last axis values as posted, pre-transform */
 } DDXTouchPointInfoRec;
 
 typedef struct _TouchClassRec {
